vllm-tt-metal-llama3-70b/src/logging_utils.py

```diff
@@ -0,0 +1,174 @@
+        )
+
+    self.ctx = zmq.Context()  # type: ignore[attr-defined]
+
+    # Receive input from the client.
+    self.input_socket = self.ctx.socket(zmq.constants.PULL)
+    self.input_socket.bind(f"{ipc_path}{IPC_INPUT_EXT}")
+
+    # Send output stream back to client.
+    self.output_socket = self.ctx.socket(zmq.constants.PUSH)
+    self.output_socket.bind(f"{ipc_path}{IPC_OUTPUT_EXT}")
+
+    # Send heartbeats back to client.
+    self.heartbeat_socket = self.ctx.socket(zmq.constants.PUSH)
+    self.heartbeat_socket.bind(f"{ipc_path}{IPC_HEALTH_EXT}")
+
+    # IPC path for the data socket.
+    self.data_ipc_path = f"{ipc_path}{IPC_DATA_EXT}"
+
+    # Error state.
+    self._errored_with: Optional[BaseException] = None
+
+    # Heartbeat thread
+    self.heartbeat_thread = threading.Thread(target=self._heartbeat_loop, daemon=True)
+    self._heartbeat_stop_event = threading.Event()
+    # The heartbeat needs to be faster than what the client will wait for
+    # The VLLM_RPC_TIMEOUT duration is in ms, and we need one in seconds
+    self.heartbeat_interval_seconds = VLLM_RPC_TIMEOUT / 5000.0
+
+    self._last_alive_time = time.time()
+    # The heartbeats can tolerate a long period of the engine chugging
+    # away at a generation request.
+    # The VLLM_RPC_TIMEOUT duration is in ms, and we need one in seconds
+    self.last_alive_threshold = VLLM_RPC_TIMEOUT * 3.0 / 1000.0
+
+
+class RawStatLogger(StatLoggerBase):
+    def __init__(self, num_scheduler_steps) -> None:
+        self.time_to_first_token = []
+        self.time_per_output_token = []
+        self.num_scheduler_steps = num_scheduler_steps
+        timestamp = datetime.now().strftime("%Y-%m-%d_%H%M%S")
+        self.filepath = f"/home/user/tests/statistics_{timestamp}.jsonl"
+        self.num_total_grouped_step = (
+            0  # number of iterations of size num_scheduler_steps
+        )
+        self.num_inference = (
+            0  # number of times inference is done (ie. how many batches)
+        )
+
+    def log(self, stats: Stats, log_to_stdout=True) -> None:
+        if len(stats.time_to_first_tokens_iter) > 0:
+            self.time_to_first_token.append(
+                stats.time_to_first_tokens_iter
+            )  # Add all values to the list
+
+            if log_to_stdout:
+                for user_idx, ttft in enumerate(stats.time_to_first_tokens_iter):
+                    logger.info(f"User {user_idx}: Time to first token {ttft:.2f} s\n")
+
+        if len(stats.time_per_output_tokens_iter) > 0:
+            tpot = [
+                time / self.num_scheduler_steps
+                for time in stats.time_per_output_tokens_iter
+            ]
+            self.time_per_output_token.append(tpot)  # Add all values to the list
+
+        self._write_to_json(stats)
+
+    def _write_to_json(self, stats):
+        data = {}
+
+        if len(stats.time_per_output_tokens_iter) > 0:
+            data["time per output token"] = {}
```

Comment:
For dictionary keys could you use snake_case e.g. `data["time to first token"]` -> `data["time_to_first_token"]` or `data["ttft"]`?

Reply:
yes, it is resolved now